tests/repository/test_remove_content.py

```diff
@@ -171,3 +173,47 @@ def test_remove_loads_units(fast_poller, requests_mocker, client):
             Unit(content_type_id="bizarre_type"),
         ]
     )
+
+
+def test_remove_with_filters(fast_poller, requests_mocker, client):
+    """Remove succeeds when given a critria/filter for removal"""
+    repo = Repository(id="some-repo")
+    repo.__dict__["_client"] = client
+
+    requests_mocker.post(
+        "https://pulp.example.com/pulp/api/v2/repositories/some-repo/actions/unassociate/",
+        [{"json": {"spawned_tasks": [{"task_id": "task1"}]}}],
+    )
+
+    requests_mocker.post(
+        "https://pulp.example.com/pulp/api/v2/tasks/search/",
+        [{"json": [{"task_id": "task1", "state": "finished"}]}],
+    )
+    filters = Criteria.and_(
+        Criteria.with_field("filename", "some.rpm"),
+        Criteria.with_field("signing_key", Matcher.in_(["key123"])),
+    )
+
+    assert repo.remove_content(
```

Comment:
Could we please have at least one case which calls remove_content with criteria and without type_ids ?

Reply:
As per the documents and the bug https://bugzilla.redhat.com/show_bug.cgi?id=1021579, type_ids should be present for the filters in criteria to be considered. If type_ids is omitted, filters are skipped and all the content is removed. I have updated the test with this assumption that type_ids will be there with the filters in the criteria.
Should there be an error thrown from _do_unassociate if there's a filter but no type_id OR the mention in the doc is fine?

Reply:
It seems a bad idea to go ahead and remove everything if there were no type_ids.  I see two options:

1. In do_unassociate, return an error if the search had no type_ids.  The error should be worded in terms of the remove_content API, e.g. "Criteria to remove_content must specify at least one unit type!"
2. If there are no type_ids, then just before submitting the search, set them to self.get_content_type_ids() (i.e. default to all the type ids)

(1) feels like the safer approach to me, as it's dangerous anyway to do a removal without specifying the type(s).

Reply:
Updated as per option(1) for both the client and fake_client